components/sup/src/manager/file_watcher.rs

```diff
@@ -1248,2217 +327,1778 @@ impl<C: Callbacks, W: Watcher> FileWatcher<C, W> {
-            DebouncedEvent::Write(ref p) | DebouncedEvent::Chmod(ref p) => {
-                match paths.get_watched_file(p) {
-                    Some(&WatchedFile::Regular(_)) => EventAction::PlainChange(p.clone()),
-                    _ => EventAction::Ignore,
-                }
-            }
-            DebouncedEvent::NoticeRemove(ref p) => Self::handle_notice_remove_event(paths, p),
-            DebouncedEvent::Remove(p) => Self::handle_remove_event(paths, p),
-            DebouncedEvent::Create(ref p) => {
-                match paths.get_watched_file(p) {
-                    None => EventAction::Ignore,
-                    Some(&WatchedFile::MissingRegular(ref c)) => {
-                        EventAction::AddRegular(c.get_paths_action_data())
+    fn handle_file_appearance(&mut self, notify_event: &Event) {
+        trace!("FileWatcher::handle_file_appearance\n| {:?}\n| {:?}\n",
+               notify_event,
+               self.watched_file);
+        for path_buf in notify_event.paths.iter() {
+            let mut should_signal = self.watched_file.path == *path_buf;
+            if !should_signal && path_buf.is_file() {
+                should_signal = Self::canonical_compare(&self.watched_file.path, path_buf)
+            };
+            trace!("| should_signal -> {:?}", should_signal);
+            if should_signal {
+                match self.manage_watch() {
+                    Ok(_) => {
+                        self.callbacks.file_appeared(&self.watched_file.path);
+                        break;
                     }
-                    // Create event for an already existing file or
-                    // directory should not happen, restart watching.
-                    Some(&WatchedFile::Regular(_)) | Some(&WatchedFile::Directory(_)) => {
-                        EventAction::RestartWatching
+                    Err(e) => {
+                        trace!("| Err -> {:?}", e);
+                        self.log_ignored_path(notify_event, path_buf, Some(e))
                     }
-                    Some(&WatchedFile::Symlink(ref c)) => {
-                        EventAction::RewireSymlink(c.get_paths_action_data())
+                }
+            } else {
+                self.log_ignored_path(notify_event, path_buf, None);
+            }
+        }
+    }
+
+    fn handle_file_modification(&mut self, notify_event: &Event) {
+        trace!("FileWatcher::handle_file_modification\n| {:?}\n| {:?}\n",
+               notify_event,
+               self.watched_file);
+        for path_buf in notify_event.paths.iter() {
+            let mut should_signal = self.watched_file.path == *path_buf;
+            if !should_signal && path_buf.is_file() {
+                should_signal = Self::canonical_compare(&self.watched_file.path, path_buf)
+            };
+            trace!("| should_signal -> {:?}", should_signal);
+            if should_signal {
+                match self.manage_watch() {
```

Comment:
if the file has been modified, is there a need to call manage_watch since the file still exists and nothing has been removed?

Reply:
The code now avoids calling manage_watch on file modification signaling.